### ISSUE TITLE
Introduce a user blockage in private and group settings

### ISSUE THREAD

**Issue by sergejsvisockis** [joejoe2/spring-chat#5]
**Business context**
To have a fully-fledged chat service it would be particularly nice to introduce a user blockage capability.
As a user, I want to be able to block people when necessary both in private and group settings as a chat admin.

Let's make it just like a real thing! ;)

**In scope**
* The user should be able to block another user
* When the user has blocked someone then blockee can neither message the blocker nor select them as a recipient
* It should be possible to block someone in a group chat settings as well
* When someone is being blocked in a group chat then they can neither message that chat nor see chat members
* Unblocking someone in both contexts would allow blockee to interact with the user and group chat just like before being blocked
* Blacklist should be available containing all the users blocked in private settings
* Group chat should contain a blacklist containing all the users blocked

**Acceptance criteria**
* An above-mentioned functionality is in place
* Proper unit and integration tests are being implemented as well
* System design diagrams are updated

**Out of scope**
Everything not mentioned in the scope

**Comment by joejoe2** [comment]
I am planning to add blockage in private just like some chat app (ex. Line). But I am wondering blockage in group because Line deos not allow that. Can you share any real experience from existing chat app ?

**Comment by sergejsvisockis** [comment]
For instance Telegram supports this.

**Comment by sergejsvisockis** [comment]
In a couple of words - it's allowed to block a group chat participants only to the chat admin. None of the other chat participants can do this.

**Comment by joejoe2** [comment]
@sergejsvisockis I have add blockage in private chat at https://github.com/joejoe2/spring-chat/pull/6. For blockage in group chat, I will make it in later weeks.

**Comment by sergejsvisockis** [comment]
@joejoe2 Perfect! No worries, that's fine :)

**Comment by sergejsvisockis** [comment]
@joejoe2 it would be also particularly valuable to describe somewhere how user and group chat blockage functionality works. Both high-level and perhaps in-depth.

**Comment by sergejsvisockis** [comment]
What is missing at the moment is the list containing all the users blocked.

As per the requirements - I as a user would like to see a full list of people I have blocked both in private same as a chat admin I would like to see those participates who are blocked.

**Comment by sergejsvisockis** [comment]
@joejoe2 What is missing at the moment is the list containing all the users blocked.

As per the requirements - I as a user would like to see a full list of people I have blocked both in private same as a chat admin I would like to see those participates who are blocked.

**Comment by joejoe2** [comment]
I am wondering whether to provide an api to get the blocked private channels(users). So I added an field for isBlocked at.

Because if there is local storage in the app/frontend, we may get the full list of channel profiles first and store it. And we can delay the channel profile update until we call the blockage api.

**Comment by sergejsvisockis** [comment]
On the Ui it's required to have a corresponding list within the profile called - "Blacklist". Upon opening this page on the UI it should perform a corresponding REST call and retrieve the list of blocked users.

For instance, I would like to be able to unblock one or several users exactly from that list. Just like in any other chat.

Let's implement it for the private chat at first to not over-complicate things. As soon as the list is done for the private chat we can move towards group settings functionality.

**Comment by joejoe2** [comment]
Ok, but there is some performance concerns in current implementation. We need a complex query to do that. Or I have to change the flag-based block check to m2m relationship.

**Comment by sergejsvisockis** [comment]
Not actually performance impact, but rather needs to re-design this functionality. It turned out to be over-complicated and non-extendable. Happens sometimes :)

That's right we need a Many-to-Many relationship here instead.

**Comment by sergejsvisockis** [comment]
Now we can proceed with the blacklist implementation. Since this one https://github.com/joejoe2/spring-chat/pull/7 is merged.

**Comment by joejoe2** [comment]
I think the "ban" is more suitable for group chat. I have looked around telegram and some other chat apps.

**Comment by sergejsvisockis** [comment]
Exactly! It might be called like this as well :)

**Comment by sergejsvisockis** [comment]
I have re-read - I'm not sure what you mean. But "ban" is applicable to both private chats (called block) and group chats. An idea behind the scenes is exactly the same. However, in the first case, it is meant to be between users - when one user blocks another.

In the case of group chats, it would prevent users from posting any messages in a group chat settings.

Both features need to be in place.

**Comment by sergejsvisockis** [comment]
> I have re-read - I'm not sure what you mean. But "ban" is applicable to both private chats (called block) and group chats. An idea behind the scenes is exactly the same. However, in the first case, it is meant to be between users - when one user blocks another to prevent unwanted interactions.
> 
> In the case of group chats, it would disallow users from posting any messages in a group chat setting potentially even to see all the messages in that group chat (depending on the chat app) same as joining that one.
> 
> Let's say in our case group chat used blockage functionality should be aligned with the following functional requirements if someone got banned:
> 
> * The user is unable to post new messages
> * The user can't see all other participant messages
> * The user is unable to re-join the chat while being under suspension

Decided that preventing the user from posting new messages within the group chat is more than enough. In case if it's required to prevent certain users to see an overall chat history then it's easier to just kick them off the chat along with the ban.

**Comment by sergejsvisockis** [comment]
After https://github.com/joejoe2/spring-chat/pull/10 merged we can turn towards the UI part.

**Comment by joejoe2** [comment]
You're right. I will start to work in two parts: private and group chat.
